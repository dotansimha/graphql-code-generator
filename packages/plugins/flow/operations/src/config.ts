import { RawDocumentsConfig } from '@graphql-codegen/visitor-plugin-common';

/**
 * @description This plugin generates Flow types based on your `GraphQLSchema` and your GraphQL operations and fragments.
 *
 * It generates types for your GraphQL documents: Query, Mutation, Subscription and Fragment.
 *
 * This plugin requires you to use `@graphql-codegen/flow` as well, because it depends on it's types.
 */
export interface FlowDocumentsPluginConfig extends RawDocumentsConfig {
  /**
   * @description Generates Flow types as Exact types.
   * @default true
   *
   * @exampleMarkdown
   * ```yml
   * generates:
   * path/to/file.ts:
   *  plugins:
   *    - flow
   *  config:
   *    useFlowExactObjects: false
   * ```
   */
  useFlowExactObjects?: boolean;
  /**
   * @description Generates read-only Flow types
   * @default false
   *
   * @exampleMarkdown
   * ```yml
   * generates:
   * path/to/file.ts:
   *  plugins:
   *    - flow
   *  config:
   *    useFlowReadOnlyTypes: true
   * ```
   */
  useFlowReadOnlyTypes?: boolean;
  /**
   * @description Flatten fragment spread and inline fragments into a simple selection set before generating.
   * @default false
   *
   * @exampleMarkdown
   * ```yml
   * generates:
   * path/to/file.ts:
   *  plugins:
   *    - typescript
   *    - typescript-operations
   *  config:
   *    flattenGeneratedTypes: true
   * ```
   */
  flattenGeneratedTypes?: boolean;
}
