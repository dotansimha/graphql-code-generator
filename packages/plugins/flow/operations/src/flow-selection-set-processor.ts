import {
  LinkField,
  PrimitiveAliasedFields,
  SelectionSetProcessorConfig,
  ProcessResult,
  BaseSelectionSetProcessor,
  indent,
  PrimitiveField,
} from '@graphql-codegen/visitor-plugin-common';
import { GraphQLObjectType, GraphQLInterfaceType } from 'graphql';

export interface FlowSelectionSetProcessorConfig extends SelectionSetProcessorConfig {
  useFlowExactObjects: boolean;
}

export class FlowWithPickSelectionSetProcessor extends BaseSelectionSetProcessor<FlowSelectionSetProcessorConfig> {
  transformAliasesPrimitiveFields(
    schemaType: GraphQLObjectType | GraphQLInterfaceType,
    fields: PrimitiveAliasedFields[]
  ): ProcessResult {
    if (fields.length === 0) {
      return [];
    }

    const useFlowExactObject = this.config.useFlowExactObjects;
    const formatNamedField = this.config.formatNamedField;
    const fieldObj = schemaType.getFields();
    const parentName =
      (this.config.namespacedImportName ? `${this.config.namespacedImportName}.` : '') +
      this.config.convertName(schemaType.name, {
        useTypesPrefix: true,
      });

    return [
      `{${useFlowExactObject ? '|' : ''} ${fields
        .map(
          aliasedField =>
            `${formatNamedField(
              aliasedField.alias,
              fieldObj[aliasedField.fieldName].type
            )}: $ElementType<${parentName}, '${aliasedField.fieldName}'>`
        )
        .join(', ')} ${useFlowExactObject ? '|' : ''}}`,
    ];
  }

  buildFieldsIntoObject(allObjectsMerged: string[]): string {
    return `...{ ${allObjectsMerged.join(', ')} }`;
  }

  buildSelectionSetFromStrings(pieces: string[]): string {
    if (pieces.length === 0) {
      return null;
    } else if (pieces.length === 1) {
      return pieces[0];
    } else {
      return `({\n  ${pieces.map(t => indent(`...${t}`)).join(`,\n`)}\n})`;
    }
  }

  transformLinkFields(fields: LinkField[]): ProcessResult {
    if (fields.length === 0) {
      return [];
    }

    const useFlowExactObject = this.config.useFlowExactObjects;
    return [
      `{${useFlowExactObject ? '|' : ''} ${fields
        .map(field => `${field.alias || field.name}: ${field.selectionSet}`)
        .join(', ')} ${useFlowExactObject ? '|' : ''}}`,
    ];
  }

  transformPrimitiveFields(
    schemaType: GraphQLObjectType | GraphQLInterfaceType,
    fields: PrimitiveField[]
  ): ProcessResult {
    if (fields.length === 0) {
      return [];
    }

    const useFlowExactObject = this.config.useFlowExactObjects;
    const formatNamedField = this.config.formatNamedField;
    const parentName =
      (this.config.namespacedImportName ? `${this.config.namespacedImportName}.` : '') +
      this.config.convertName(schemaType.name, {
        useTypesPrefix: true,
      });
    const fieldObj = schemaType.getFields();
    let hasConditionals = false;
    const conditilnalsList: string[] = [];
    let resString = `$Pick<${parentName}, {${useFlowExactObject ? '|' : ''} ${fields
      .map(field => {
        if (field.isConditional) {
          hasConditionals = true;
          conditilnalsList.push(field.fieldName);
        }
        return `${formatNamedField(field.fieldName, fieldObj[field.fieldName].type)}: *`;
      })
      .join(', ')} ${useFlowExactObject ? '|' : ''}}>`;
    if (hasConditionals) {
      resString = `$MakeOptional<${resString}, ${conditilnalsList.map(field => `{ ${field}: * }`).join(' | ')}>`;
    }
    return [resString];
  }

  transformTypenameField(type: string, name: string): ProcessResult {
    return [`{ ${name}: ${type} }`];
  }
}
