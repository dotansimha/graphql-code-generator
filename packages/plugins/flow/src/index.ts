import { DocumentFile, PluginFunction } from 'graphql-codegen-core';
import { parse, printSchema, visit, GraphQLSchema } from 'graphql';
import { FlowVisitor } from './visitor';
import { RawConfig } from 'graphql-codegen-visitor-plugin-common';
export { DEFAULT_SCALARS } from 'graphql-codegen-visitor-plugin-common/src/scalars';
export * from 'graphql-codegen-visitor-plugin-common/src/variables-to-object';
export * from './visitor';

export interface FlowPluginConfig extends RawConfig {
  useFlowExactObjects?: boolean;
  useFlowReadOnlyTypes?: boolean;
}

export const plugin: PluginFunction<FlowPluginConfig> = (
  schema: GraphQLSchema,
  documents: DocumentFile[],
  config: FlowPluginConfig
) => {
  const result = `/* @flow */\n\n`;
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);

  const visitorResult = visit(astNode, {
    leave: new FlowVisitor(config)
  });

  return result + visitorResult.definitions.join('\n');
};
