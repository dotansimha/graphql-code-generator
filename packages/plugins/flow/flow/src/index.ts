import { Types, PluginFunction } from '@graphql-codegen/plugin-helpers';
import { parse, printSchema, visit, GraphQLSchema } from 'graphql';
import { FlowVisitor } from './visitor';
import { FlowPluginConfig } from './config';

export * from './visitor';
export * from './flow-variables-to-object';

export const plugin: PluginFunction<FlowPluginConfig, Types.ComplexPluginOutput> = (
  schema: GraphQLSchema,
  documents: Types.DocumentFile[],
  config: FlowPluginConfig
) => {
  const header = `// @flow\n`;
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);
  const visitor = new FlowVisitor(schema, config);

  const visitorResult = visit(astNode, {
    leave: visitor,
  });

  return {
    prepend: [header, ...visitor.getEnumsImports()],
    content: [visitor.scalarsDefinition, ...visitorResult.definitions].join('\n'),
  };
};
