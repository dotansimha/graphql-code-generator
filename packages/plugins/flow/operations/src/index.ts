import { PluginFunction, Types } from '@graphql-codegen/plugin-helpers';
import { visit, concatAST, GraphQLSchema, Kind, FragmentDefinitionNode } from 'graphql';
import { FlowDocumentsVisitor } from './visitor';
import { LoadedFragment, optimizeOperations } from '@graphql-codegen/visitor-plugin-common';
import { FlowDocumentsPluginConfig } from './config';

export const plugin: PluginFunction<FlowDocumentsPluginConfig> = (
  schema: GraphQLSchema,
  rawDocuments: Types.DocumentFile[],
  config: FlowDocumentsPluginConfig
) => {
  const documents = config.flattenGeneratedTypes
    ? optimizeOperations(schema, rawDocuments, { includeFragments: true })
    : rawDocuments;

  const prefix = config.preResolveTypes
    ? ''
    : `type $Pick<Origin: Object, Keys: Object> = $ObjMapi<Keys, <Key>(k: Key) => $ElementType<Origin, Key>>;\n`;

  const allAst = concatAST(documents.map(v => v.document));
  const includedFragments = allAst.definitions.filter(d => d.kind === Kind.FRAGMENT_DEFINITION);

  const allFragments: LoadedFragment[] = [
    ...(includedFragments as FragmentDefinitionNode[]).map(fragmentDef => ({
      node: fragmentDef,
      name: fragmentDef.name.value,
      onType: fragmentDef.typeCondition.name.value,
      isExternal: false,
    })),
    ...(config.externalFragments || []),
  ];

  const visitorResult = visit(allAst, {
    leave: new FlowDocumentsVisitor(schema, config, allFragments),
  });

  return {
    prepend: ['// @flow \n'],
    content: [prefix, ...visitorResult.definitions].join('\n'),
  };
};
