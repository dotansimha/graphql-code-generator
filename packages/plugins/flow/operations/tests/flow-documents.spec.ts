import { Types } from '@graphql-codegen/plugin-helpers';
import '@graphql-codegen/testing';
import { parse, buildClientSchema, buildSchema } from 'graphql';
import { plugin } from '../src/index';
import { validateFlow } from '../../flow/tests/validate-flow';
import { readFileSync } from 'fs';

describe('Flow Operations Plugin', () => {
  const gitHuntSchema = buildClientSchema(JSON.parse(readFileSync('../../../../dev-test/githunt/schema.json', 'utf-8')));
  const schema = buildSchema(/* GraphQL */ `
    type User {
      id: ID!
      username: String!
      email: String!
      profile: Profile
      role: Role
    }

    type Profile {
      age: Int
      firstName: String!
    }

    type Mutation {
      login(username: String!, password: String!): User
    }

    type Subscription {
      userCreated: User
    }

    interface Notifiction {
      id: ID!
    }

    type TextNotification implements Notifiction {
      id: ID!
      text: String!
    }

    type ImageNotification implements Notifiction {
      id: ID!
      imageUrl: String!
      metadata: ImageMetadata!
    }

    type ImageMetadata {
      createdBy: String!
    }

    enum Role {
      USER
      ADMIN
    }

    union MyUnion = User | Profile

    type Query {
      me: User
      unionTest: MyUnion
      notifications: [Notifiction!]!
      dummy: String
      dummyNonNull: String!
      dummyArray: [String]
      dummyNonNullArray: [String]!
      dummyNonNullArrayWithValues: [String!]!
      dummyWithType: Profile
    }

    schema {
      query: Query
      mutation: Mutation
      subscription: Subscription
    }
  `);

  describe('Naming Convention & Types Prefix', () => {
    it('Should allow custom naming and point to the correct type', async () => {
      const ast = parse(/* GraphQL */ `
        query notifications {
          notifications {
            id

            ... on TextNotification {
              text
            }

            ... on ImageNotification {
              imageUrl
              metadata {
                createdBy
              }
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { namingConvention: 'lower-case#lowerCase' },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type notificationsquery = ({
        ...{ __typename?: 'Query' },
      ...{| notifications: Array<({
          ...{ __typename?: 'TextNotification' },
        ...$Pick<textnotification, {| text: *, id: * |}>
      }) | ({
          ...{ __typename?: 'ImageNotification' },
        ...$Pick<imagenotification, {| imageUrl: *, id: * |}>,
        ...{| metadata: ({
            ...{ __typename?: 'ImageMetadata' },
          ...$Pick<imagemetadata, {| createdBy: * |}>
        }) |}
      })> |}
    });
      `);
      validateFlow(result);
    });

    it('Should allow custom naming and point to the correct type - with custom prefix', async () => {
      const ast = parse(/* GraphQL */ `
        query notifications {
          notifications {
            id

            ... on TextNotification {
              text
            }

            ... on ImageNotification {
              imageUrl
              metadata {
                createdBy
              }
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { typesPrefix: 'i', namingConvention: 'lower-case#lowerCase' },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`export type inotificationsqueryvariables = {};`);
      expect(result).toBeSimilarStringTo(`
      export type inotificationsquery = ({
        ...{ __typename?: 'Query' },
      ...{| notifications: Array<({
          ...{ __typename?: 'TextNotification' },
        ...$Pick<itextnotification, {| text: *, id: * |}>
      }) | ({
          ...{ __typename?: 'ImageNotification' },
        ...$Pick<iimagenotification, {| imageUrl: *, id: * |}>,
        ...{| metadata: ({
            ...{ __typename?: 'ImageMetadata' },
          ...$Pick<iimagemetadata, {| createdBy: * |}>
        }) |}
      })> |}
    });
      `);
      validateFlow(result);
    });
  });

  describe('Import namespace', () => {
    it('Should handle "namespacedImportName" and add it when specified', async () => {
      const ast = parse(/* GraphQL */ `
        query notifications {
          notifications {
            id

            ... on TextNotification {
              text
            }

            ... on ImageNotification {
              imageUrl
              metadata {
                createdBy
              }
            }
          }
        }
      `);
      const result = await plugin(schema, [{ filePath: '', content: ast }], { namespacedImportName: 'Types' }, { outputFile: '' });

      expect(result).toBeSimilarStringTo(`
      export type NotificationsQuery = ({
        ...{ __typename?: 'Query' },
      ...{| notifications: Array<({
          ...{ __typename?: 'TextNotification' },
        ...$Pick<Types.TextNotification, {| text: *, id: * |}>
      }) | ({
          ...{ __typename?: 'ImageNotification' },
        ...$Pick<Types.ImageNotification, {| imageUrl: *, id: * |}>,
        ...{| metadata: ({
            ...{ __typename?: 'ImageMetadata' },
          ...$Pick<Types.ImageMetadata, {| createdBy: * |}>
        }) |}
      })> |}
    });
      `);
      validateFlow(result);
    });
  });

  describe('__typename', () => {
    it('Should skip __typename when skipTypename is set to true', async () => {
      const ast = parse(`
        query {
          dummy
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );
      expect(result).not.toContain(`__typename`);
      validateFlow(result);
    });

    it('Should add __typename as non-optional when explicitly specified', async () => {
      const ast = parse(/* GraphQL */ `
        query {
          __typename
          dummy
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        {},
        { outputFile: '' }
      );
      expect(result).toBeSimilarStringTo(`
      export type Unnamed_1_Query = ({
        ...{ __typename: 'Query' },
      ...$Pick<Query, {| dummy: * |}>
    });
      `);
      validateFlow(result);
    });

    it('Should add __typename as optional when its not specified', async () => {
      const ast = parse(/* GraphQL */ `
        query {
          dummy
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        {},
        { outputFile: '' }
      );
      expect(result).toBeSimilarStringTo(`
      export type Unnamed_1_Query = ({
        ...{ __typename?: 'Query' },
      ...$Pick<Query, {| dummy: * |}>
    });
      `);
      validateFlow(result);
    });

    it('Should add __typename as non-optional when its explictly specified, even if skipTypename is true', async () => {
      const ast = parse(/* GraphQL */ `
        query {
          __typename
          dummy
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );
      expect(result).toBeSimilarStringTo(`
      export type Unnamed_1_Query = ({
        ...{ __typename: 'Query' },
      ...$Pick<Query, {| dummy: * |}>
    });
      `);
      validateFlow(result);
    });

    it('Should add __typename correctly when unions are in use', async () => {
      const ast = parse(/* GraphQL */ `
        query unionTest {
          unionTest {
            ... on User {
              id
            }

            ... on Profile {
              age
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        {},
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type UnionTestQuery = ({
        ...{ __typename?: 'Query' },
      ...{| unionTest: ?({
          ...{ __typename?: 'User' },
        ...$Pick<User, {| id: * |}>
      }) | ({
          ...{ __typename?: 'Profile' },
        ...$Pick<Profile, {| age: * |}>
      }) |}
    });
      `);
      validateFlow(result);
    });

    it('Should add non optional __typename when specified in config', async () => {
      const ast = parse(/* GraphQL */ `
        query unionTest {
          unionTest {
            ... on User {
              id
            }

            ... on Profile {
              age
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { nonOptionalTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type UnionTestQuery = ({
        ...{ __typename: 'Query' },
      ...{| unionTest: ?({
          ...{ __typename: 'User' },
        ...$Pick<User, {| id: * |}>
      }) | ({
          ...{ __typename: 'Profile' },
        ...$Pick<Profile, {| age: * |}>
      }) |}
    });
      `);
      validateFlow(result);
    });

    it('Should add __typename correctly when interfaces are in use', async () => {
      const ast = parse(/* GraphQL */ `
        query notifications {
          notifications {
            id

            ... on TextNotification {
              text
            }

            ... on ImageNotification {
              imageUrl
              metadata {
                createdBy
              }
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        {},
        { outputFile: '' }
      );
      expect(result).toBeSimilarStringTo(`
      export type NotificationsQuery = ({
        ...{ __typename?: 'Query' },
      ...{| notifications: Array<({
          ...{ __typename?: 'TextNotification' },
        ...$Pick<TextNotification, {| text: *, id: * |}>
      }) | ({
          ...{ __typename?: 'ImageNotification' },
        ...$Pick<ImageNotification, {| imageUrl: *, id: * |}>,
        ...{| metadata: ({
            ...{ __typename?: 'ImageMetadata' },
          ...$Pick<ImageMetadata, {| createdBy: * |}>
        }) |}
      })> |}
    });
      `);
      validateFlow(result);
    });
  });

  describe('Unnamed Documents', () => {
    it('Should handle unnamed documents correctly', async () => {
      const ast = parse(`
        query {
          dummy
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`export type Unnamed_1_Query = $Pick<Query, {| dummy: * |}>;`);
      expect(result).toBeSimilarStringTo(`export type Unnamed_1_QueryVariables = {};`);
      validateFlow(result);
    });

    it('Should handle unnamed documents correctly with multiple documents', async () => {
      const ast = parse(`
        query {
          dummy
        }

        query {
          dummy
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );
      expect(result).toBeSimilarStringTo(`export type Unnamed_1_Query = $Pick<Query, {| dummy: * |}>;`);
      expect(result).toBeSimilarStringTo(`export type Unnamed_1_QueryVariables = {};`);
      expect(result).toBeSimilarStringTo(`export type Unnamed_2_Query = $Pick<Query, {| dummy: * |}>;`);
      expect(result).toBeSimilarStringTo(`export type Unnamed_2_QueryVariables = {};`);
      validateFlow(result);
    });
  });

  describe('Selection Set', () => {
    it('Should support fragment spread correctly with simple type with no other fields', async () => {
      const ast = parse(/* GraphQL */ `
        fragment UserFields on User {
          id
          username
          profile {
            age
          }
          role
        }

        query me {
          me {
            ...UserFields
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type MeQuery = {| me: ?UserFieldsFragment |};
      `);
      validateFlow(result);
    });

    it('Should support fragment spread correctly with simple type with other fields', async () => {
      const ast = parse(/* GraphQL */ `
        fragment UserFields on User {
          id
          profile {
            age
          }
        }

        query me {
          me {
            ...UserFields
            username
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type MeQuery = {| me: ?({
        ...$Pick<User, {| username: * |}>,
      ...UserFieldsFragment
    }) |};
      `);
      validateFlow(result);
    });

    it('Should support fragment spread correctly with multiple fragment spread', async () => {
      const ast = parse(/* GraphQL */ `
        fragment UserFields on User {
          id
        }

        fragment UserProfile on User {
          profile {
            age
          }
        }

        query me {
          me {
            ...UserFields
            ...UserProfile
            username
          }
        }
      `);

      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type MeQuery = {| me: ?({
        ...$Pick<User, {| username: * |}>,
      ...UserFieldsFragment,
      ...UserProfileFragment
    }) |};
      `);
      validateFlow(result);
    });

    it('Should support interfaces correctly when used with inline fragments', async () => {
      const ast = parse(/* GraphQL */ `
        query notifications {
          notifications {
            id

            ... on TextNotification {
              text
            }

            ... on ImageNotification {
              imageUrl
              metadata {
                createdBy
              }
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type NotificationsQuery = {| notifications: Array<$Pick<TextNotification, {| text: *, id: * |}> | ({
        ...$Pick<ImageNotification, {| imageUrl: *, id: * |}>,
      ...{| metadata: $Pick<ImageMetadata, {| createdBy: * |}> |}
    })> |};
      `);
      validateFlow(result);
    });

    it('Should support union correctly when used with inline fragments', async () => {
      const ast = parse(/* GraphQL */ `
        query unionTest {
          unionTest {
            ... on User {
              id
            }

            ... on Profile {
              age
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type UnionTestQuery = {| unionTest: ?$Pick<User, {| id: * |}> | $Pick<Profile, {| age: * |}> |};
      `);
      validateFlow(result);
    });

    it('Should support inline fragments', async () => {
      const ast = parse(/* GraphQL */ `
        query currentUser {
          me {
            id
            ... on User {
              username
              profile {
                age
              }
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type CurrentUserQuery = {| me: ?({
        ...$Pick<User, {| username: *, id: * |}>,
      ...{| profile: ?$Pick<Profile, {| age: * |}> |}
    }) |};
      `);
      validateFlow(result);
    });

    it('Should build a basic selection set based on basic query on GitHub schema', async () => {
      const ast = parse(/* GraphQL */ `
        query me($repoFullName: String!) {
          currentUser {
            login
            html_url
          }
          entry(repoFullName: $repoFullName) {
            id
            postedBy {
              login
              html_url
            }
            createdAt
          }
        }
      `);
      const result = await plugin(
        gitHuntSchema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );
      expect(result).toBeSimilarStringTo(
        `export type MeQueryVariables = {
          repoFullName: $ElementType<Scalars, 'String'>
        };`
      );
      expect(result).toBeSimilarStringTo(`
      export type MeQuery = {| currentUser: ?$Pick<User, {| login: *, html_url: * |}>, entry: ?({
        ...$Pick<Entry, {| id: *, createdAt: * |}>,
      ...{| postedBy: $Pick<User, {| login: *, html_url: * |}> |}
    }) |};
      `);
      validateFlow(result);
    });

    it('Should build a basic selection set based on basic query', async () => {
      const ast = parse(/* GraphQL */ `
        query dummy {
          dummy
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );
      expect(result).toBeSimilarStringTo(`export type DummyQuery = $Pick<Query, {| dummy: * |}>;`);
      validateFlow(result);
    });

    it('Should build a basic selection set based on basic query with field aliasing for basic scalar', async () => {
      const ast = parse(/* GraphQL */ `
        query dummy {
          customName: dummy
          customName2: dummyWithType {
            age
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type DummyQuery = ({
        ...{| customName: $ElementType<Query, 'dummy'> |},
      ...{| customName2: ?$Pick<Profile, {| age: * |}> |}
    });
      `);
      validateFlow(result);
    });

    it('Should build a basic selection set based on a query with inner fields', async () => {
      const ast = parse(/* GraphQL */ `
        query currentUser {
          me {
            id
            username
            role
            profile {
              age
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type CurrentUserQuery = {| me: ?({
        ...$Pick<User, {| id: *, username: *, role: * |}>,
      ...{| profile: ?$Pick<Profile, {| age: * |}> |}
    }) |};
      `);

      validateFlow(result);
    });
  });

  describe('Fragment Definition', () => {
    it('Should build fragment definition correctly - with name and selection set', async () => {
      const ast = parse(/* GraphQL */ `
        fragment UserFields on User {
          id
          username
          profile {
            age
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type UserFieldsFragment = ({
        ...$Pick<User, {| id: *, username: * |}>,
      ...{| profile: ?$Pick<Profile, {| age: * |}> |}
    });
      `);
      validateFlow(result);
    });
  });

  describe('Operation Definition', () => {
    it('Should detect Mutation correctly', async () => {
      const ast = parse(/* GraphQL */ `
        mutation login {
          login(username: "1", password: "2") {
            id
            username
            profile {
              age
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type LoginMutation = {| login: ?({
        ...$Pick<User, {| id: *, username: * |}>,
      ...{| profile: ?$Pick<Profile, {| age: * |}> |}
    }) |};
      `);
      validateFlow(result);
    });

    it('Should detect Query correctly', async () => {
      const ast = parse(`
        query test {
          dummy
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`export type TestQuery = $Pick<Query, {| dummy: * |}>;`);
      validateFlow(result);
    });

    it('Should detect Subscription correctly', async () => {
      const ast = parse(`
        subscription test {
          userCreated {
            id
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );
      expect(result).toBeSimilarStringTo(`export type TestSubscription = {| userCreated: ?$Pick<User, {| id: * |}> |};`);
      validateFlow(result);
    });

    it('Should handle operation variables correctly', async () => {
      const ast = parse(`
        query testQuery($username: String, $email: String, $password: String!, $input: InputType, $mandatoryInput: InputType!, $testArray: [String], $requireString: [String]!, $innerRequired: [String!]!) {
          dummy
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(
        `export type TestQueryQueryVariables = {
          username?: ?$ElementType<Scalars, 'String'>,
          email?: ?$ElementType<Scalars, 'String'>,
          password: $ElementType<Scalars, 'String'>,
          input?: ?InputType,
          mandatoryInput: InputType,
          testArray?: ?Array<?$ElementType<Scalars, 'String'>>,
          requireString: Array<?$ElementType<Scalars, 'String'>>,
          innerRequired: Array<$ElementType<Scalars, 'String'>>
        };`
      );
      validateFlow(result);
    });

    it('Should create empty variables when there are no operation variables', async () => {
      const ast = parse(`
        query testQuery {
          dummy
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`export type TestQueryQueryVariables = {};`);
      validateFlow(result);
    });
  });

  describe('Output options', () => {
    it('Should respect flow option useFlowExactObjects=false', async () => {
      const ast = parse(/* GraphQL */ `
        query currentUser {
          me {
            id
            username
            role
            profile {
              age
            }
          }
        }
      `);
      const result = await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true, useFlowExactObjects: false },
        { outputFile: '' }
      );

      expect(result).toBeSimilarStringTo(`
      export type CurrentUserQuery = { me: ?({
        ...$Pick<User, { id: *, username: *, role: * }>,
      ...{ profile: ?$Pick<Profile, { age: * }> }
    }) };
      `);

      validateFlow(result);
    });

    it('Should respect flow option useFlowReadOnlyTypes', async () => {
      const ast = parse(/* GraphQL */ `
        query currentUser {
          me {
            id
            username
            role
            profile {
              age
            }
          }
        }
      `);
      const result = (await plugin(
        schema,
        [
          {
            filePath: '',
            content: ast,
          },
        ],
        { skipTypename: true, useFlowReadOnlyTypes: true },
        { outputFile: '' }
      )) as Types.ComplexPluginOutput;
      expect(result.content).toMatchSnapshot();
      expect(result).toBeSimilarStringTo(`
      export type CurrentUserQuery = {| +me: ?({
        ...$Pick<User, {| +id: *, +username: *, +role: * |}>,
      ...{| +profile: ?$Pick<Profile, {| +age: * |}> |}
    }) |};
      `);

      validateFlow(result);
    });
  });
});
