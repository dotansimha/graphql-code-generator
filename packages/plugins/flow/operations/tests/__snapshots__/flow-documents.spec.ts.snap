// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flow Operations Plugin Output options Should respect flow option useFlowReadOnlyTypes 1`] = `
"type $Pick<Origin: Object, Keys: Object> = $ObjMapi<Keys, <Key>(k: Key) => $ElementType<Origin, Key>>;

export type CurrentUserQueryVariables = {};


export type CurrentUserQuery = {| +me?: ?({
      ...$Pick<User, {| +id: *, +username: *, +role?: * |}>,
    ...{| +profile?: ?$Pick<Profile, {| +age?: * |}> |}
  }) |};
"
`;
